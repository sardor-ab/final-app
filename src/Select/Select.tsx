import React, { createRef, PureComponent } from "react";
import cn from "classnames";
import { Input } from "../components/Input";
import { filterList } from "utils";
import { AbsolutList } from "./components/list/absolutList";
import { FixedList } from "./components/list/fixedList";
import * as css from "./select.module.scss";

export interface ISelectItem {
  [key: string]: any;
}

export type ISelectChange = (item: any, name?: string) => void;
export type ISelectClear = (item: null) => void;

interface IProps {
  "data-test-id"?: string;
  style?: React.CSSProperties;
  isLoading?: boolean;
  disabled?: boolean;
  required?: boolean;
  fixedPosition?: boolean;
  name?: string;
  label?: string;
  className?: string;
  helperText?: string;
  placeholder?: string;
  selectedId?: string | number | boolean;
  items?: ISelectItem[];
  onChange: ISelectChange;
  onClear?: ISelectClear;
}

interface IState {
  isOpen: boolean;
  search: string;
}

export class Select extends PureComponent<IProps, IState> {
  inputRef = createRef<HTMLInputElement>();

  state = {
    isOpen: false,
    search: "",
  };

  onFocus = () => {
    this.setState({
      isOpen: true,
      search: "",
    });
  };

  onClose = () => {
    // HTMLInputElement
    const inputNode: any = this.inputRef.current;

    inputNode.blur();
  };

  onBlur = () => {
    this.setState({
      isOpen: false,
    });
  };

  onClear = () => {
    const { onClear } = this.props;

    this.setState(
      {
        isOpen: false,
        search: "",
      },
      () => onClear(null)
    );
  };

  handleInputChange = ({ target: { value } }: any) => {
    this.setState({
      search: value,
    });
  };

  handleSelected = (item: ISelectItem) => {
    const { name, onChange } = this.props;

    onChange(item, name);
  };

  renderList = () => {
    const { selectedId, isLoading, items, fixedPosition } = this.props;
    const { search } = this.state;
    const list = filterList(items || [], search);

    if (fixedPosition) {
      const inputPosition =
        this.inputRef.current?.getBoundingClientRect() || null;

      return (
        <FixedList
          inputPosition={inputPosition}
          isLoading={Boolean(isLoading)}
          selectedId={selectedId || null}
          onSelected={this.handleSelected}
          onClose={this.onClose}
          list={list}
          items={items || []}
        />
      );
    }

    return (
      <AbsolutList
        isLoading={Boolean(isLoading)}
        selectedId={selectedId || null}
        onSelected={this.handleSelected}
        onClose={this.onClose}
        list={list}
        items={items || []}
      />
    );
  };

  render() {
    const {
      placeholder,
      className,
      disabled,
      selectedId,
      helperText,
      label,
      isLoading,
      required,
      style,
      items,
      onClear,
    } = this.props;
    const { isOpen, search } = this.state;
    const selectedValue = items?.find((item) => item.id === selectedId)?.name;
    // eslint-disable-next-line react/destructuring-assignment
    const dataTestId = this.props["data-test-id"] || "select";

    return (
      <div
        data-test-id={dataTestId}
        style={style}
        className={cn(css.wrapper, className)}
      >
        <div className={css.select}>
          <Input
            data-test-id={`${dataTestId}_input`}
            label={label}
            isLoading={isLoading}
            id={dataTestId}
            inputRef={this.inputRef}
            className={cn(css.input, isOpen && css.open)}
            required={required}
            disabled={disabled}
            placeholder={selectedValue || placeholder || ""}
            value={isOpen ? search : selectedValue}
            onChange={this.handleInputChange}
            onFocus={this.onFocus}
            onBlur={this.onBlur}
            onClear={onClear && this.onClear}
            helperText={helperText}
          />
          <div
            className={cn(
              css.arrow,
              isOpen && css.arrowOpen,
              onClear && css.arrowWithClear
            )}
          >
            <span>&lsaquo;</span>
          </div>
          {isOpen && this.renderList()}
        </div>
      </div>
    );
  }
}
