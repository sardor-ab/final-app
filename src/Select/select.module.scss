.wrapper {
	width: 100%;
}

.select {
	position: relative;
	width: 100%;

	p {
		bottom: -17px;
		overflow: hidden;
		position: absolute;
		text-overflow: ellipsis;
		user-select: none;
		white-space: nowrap;
		width: 100%;
	}
}

.inputWrapper {
	display: flex;
	flex-direction: column;
	position: relative;
}

.input input {
	cursor: pointer;
    padding-left: 10px;
	padding-right: 24px;
	text-overflow: ellipsis;
}

.open input {
	cursor: auto;
}

.arrow {
    position: absolute;
    right: 10px;
    top: 10px;
    bottom: 0;
    transform: rotate(270deg);
    transition: transform 300ms ease-in-out;
    width: 24px;
    height: 24px;
    pointer-events: none;
    user-select: none;
    border-radius: 50%;

	span {
        position: absolute;
        color: #a0a0a0;
        font-size: 38px;
        left: 6px;
        top: -22px;
	}
}

.arrowOpen {
	cursor: pointer;
	pointer-events: auto;
	transform: rotate(90deg);
}

.arrowWithClear {
    right: 30px;
}
