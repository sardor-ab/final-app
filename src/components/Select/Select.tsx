import React, { createRef, PureComponent } from "react";
import cn from "classnames";
import { Input } from "../Input";
import filterList from "../../utils";
import { AbsolutList } from "./components/list/absolutList";
import { FixedList } from "./components/list/fixedList";
import css from "./select.module.scss";

export interface ISelectItem {
  [key: string]: any;
}

export type ISelectChange = (
  item: any,
  multiple: boolean,
  name?: string
) => void;
export type ISelectClear = (item: null, multiple: boolean) => void;

interface IProps {
  "data-test-id"?: string;
  style?: React.CSSProperties;
  isLoading?: boolean;
  disabled?: boolean;
  required?: boolean;
  fixedPosition?: boolean;
  name?: string;
  label?: string;
  className?: string;
  helperText?: string;
  placeholder?: string;
  selectedId?: string | number | boolean;
  multiple: boolean;
  items?: ISelectItem[];
  onChange: ISelectChange;
  onClear?: ISelectClear;
}

interface IState {
  isOpen: boolean;
  search: string;
}

export class Select extends PureComponent<IProps, IState> {
  inputRef = createRef<HTMLInputElement>();

  state = {
    isOpen: false,
    search: "",
  };

  onFocus = () => {
    this.setState({
      isOpen: true,
      search: "",
    });
  };

  onClose = () => {
    // HTMLInputElement
    const inputNode: any = this.inputRef.current;

    inputNode.blur();
  };

  onBlur = () => {
    this.setState({
      isOpen: false,
    });
  };

  onClear = () => {
    const { onClear, multiple } = this.props;

    this.setState(
      {
        isOpen: false,
        search: "",
      },
      // () => onClear(null);
      () => onClear?.(null, multiple)
    );
  };

  handleInputChange = ({ target: { value } }: any) => {
    this.setState({
      search: value,
    });
  };

  handleSelected = (item: ISelectItem) => {
    const { name, onChange, multiple } = this.props;

    onChange(item, multiple, name);
  };

  renderList = () => {
    const { selectedId, isLoading, items, fixedPosition } = this.props;
    const { search } = this.state;
    const list = filterList(items || [], search);

    if (fixedPosition) {
      const inputPosition =
        this.inputRef.current?.getBoundingClientRect() || null;

      return (
        <FixedList
          inputPosition={inputPosition}
          isLoading={Boolean(isLoading)}
          selectedId={selectedId || null}
          onSelected={this.handleSelected}
          onClose={this.onClose}
          list={list}
          items={items || []}
        />
      );
    }

    return (
      <AbsolutList
        isLoading={Boolean(isLoading)}
        selectedId={selectedId || null}
        onSelected={this.handleSelected}
        onClose={this.onClose}
        list={list}
        items={items || []}
      />
    );
  };

  render() {
    const {
      placeholder,
      className,
      disabled,
      selectedId,
      helperText,
      label,
      isLoading,
      required,
      style,
      items,
      multiple,
      onClear,
    } = this.props;

    const { isOpen, search } = this.state;
    const selectedValue = multiple
      ? items
          ?.filter((item) => selectedId?.toString().includes(item.id))
          .map((i) => i.name)
          .join(", ")
      : items?.find((item) => item.id === selectedId)?.name;
    // eslint-disable-next-line react/destructuring-assignment
    const dataTestId = this.props["data-test-id"] || "select";

    return (
      <div
        data-test-id={dataTestId}
        style={style}
        className={cn(css.wrapper, className)}
      >
        <div className={css.select}>
          <Input
            data-test-id={`${dataTestId}_input`}
            label={label}
            isLoading={isLoading}
            id={dataTestId}
            inputRef={this.inputRef}
            className={cn(css.input, isOpen && css.open)}
            required={required}
            disabled={disabled}
            placeholder={selectedValue || placeholder || ""}
            value={isOpen ? search : selectedValue}
            onChange={this.handleInputChange}
            onFocus={this.onFocus}
            onBlur={this.onBlur}
            onClear={onClear && this.onClear}
            helperText={helperText}
          />
          <div
            role="presentation"
            onClick={this.onBlur}
            className={cn(
              css.arrow,
              isOpen && css.arrowOpen,
              onClear && css.arrowWithClear
            )}
          >
            <span>&lsaquo;</span>
          </div>
          {isOpen && this.renderList()}
        </div>
      </div>
    );
  }
}
