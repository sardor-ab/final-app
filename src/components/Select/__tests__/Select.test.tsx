import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { Select } from '../Select';

describe('Select', () => {
    const items = [
        { id: 'id_1', name: 'name_1' },
        { id: 'id_2', name: 'name_2' },
    ];

    test('Снимок компонента', () => {
        const { getByTestId } = render(<Select items={items} onChange={jest.fn()} />);

        expect(getByTestId('select')).toMatchSnapshot();
    });

    test('Снимок компонента с props.fixedPosition и видимым выпадающем списком', () => {
        const { getByTestId } = render(<Select fixedPosition items={items} onChange={jest.fn()} />);

        getByTestId('select_input').focus();

        expect(getByTestId('select')).toMatchSnapshot();
    });

    test('Компонента с props.fixedPosition и "Ничего не найдено"', () => {
        const { getByTestId, queryByText } = render(
            <Select fixedPosition items={items} onChange={jest.fn()} />
        );

        getByTestId('select_input').focus();

        fireEvent.change(getByTestId('select_input'), { target: { value: 'name_2222' } });

        expect(queryByText(/Ничего не найдено/)).toBeInTheDocument();
    });

    test('Компонента с props.fixedPosition и "нет данных"', () => {
        const { getByTestId, queryByText } = render(
            <Select fixedPosition items={undefined} onChange={jest.fn()} />
        );

        getByTestId('select_input').focus();

        expect(queryByText(/Нет данных/)).toBeInTheDocument();
    });

    test('Компонент с props.fixedPosition=false и "нет данных"', () => {
        const { getByTestId, queryByText } = render(
            <Select fixedPosition={false} items={undefined} onChange={jest.fn()} />
        );

        getByTestId('select_input').focus();

        expect(queryByText(/Нет данных/)).toBeInTheDocument();
    });

    // eslint-disable-next-line max-len
    test('Снимок компонента с props.fixedPosition=false (position: absolut) и видимым выпадающем списком', () => {
        const { getByTestId } = render(
            <Select fixedPosition={false} items={items} onChange={jest.fn()} />
        );

        getByTestId('select_input').focus();

        expect(getByTestId('select')).toMatchSnapshot();
    });

    test('Если передать props.isLoading, то в DOM должен появиться Loader', () => {
        const { getByTestId } = render(<Select items={items} isLoading onChange={jest.fn()} />);

        expect(getByTestId(/loader/)).toBeInTheDocument();
    });

    test('Если передать props.className, то оно должно появиться в DOM', () => {
        const className = 'amazingClassName';
        const { getByTestId } = render(
            <Select items={items} className={className} onChange={jest.fn()} />
        );

        expect(getByTestId('select')).toHaveClass(className);
    });

    test('Если передать props.style, то в DOM должна появиться inline стиль для компонента', () => {
        const style = { background: 'red' };
        const { getByTestId } = render(<Select style={style} items={items} onChange={jest.fn()} />);

        expect(getByTestId('select')).toHaveStyle(style);
    });

    // eslint-disable-next-line max-len
    test('Если передать props.isLoading и выбрать input, то в DOM должна появиться надпись "...загрузка"', () => {
        const { getByText, getByTestId } = render(
            <Select items={items} isLoading onChange={jest.fn()} />
        );

        getByTestId('select_input').focus();

        expect(getByText(/...загрузка/)).toBeInTheDocument();
    });

    // eslint-disable-next-line max-len
    test('Если передать props.items пустой массив и выбрать input, то в DOM должна появиться надпись "Нет данных"', () => {
        const { getByText, getByTestId } = render(<Select items={[]} onChange={jest.fn()} />);

        getByTestId('select_input').focus();

        expect(getByText(/Нет данных/)).toBeInTheDocument();
    });

    test('Если передать props.disabled, то выпадающий список не должен появляться', () => {
        const { getByTestId, queryByText } = render(
            <Select disabled items={items} onChange={jest.fn()} />
        );

        getByTestId('select_input').focus();

        expect(queryByText(/name_1/)).not.toBeInTheDocument();
    });

    test('Если передать props.label, то его значение должно появиться в DOM', () => {
        const { getByText } = render(
            <Select label="label" required items={items} onChange={jest.fn()} />
        );

        expect(getByText(/label/)).toBeInTheDocument();
    });

    // eslint-disable-next-line max-len
    test('Если передать props.required и props.label, то должна в DOM появиться label с "*"', () => {
        const { getByText } = render(
            <Select label="label" required items={items} onChange={jest.fn()} />
        );

        expect(getByText(/label */)).toBeInTheDocument();
    });

    // eslint-disable-next-line max-len
    test('Если передать props.helperText, то должна в DOM появиться компонент HelperText с содержимым', () => {
        const helperText = 'amazingHelperText';
        const { getByText, getByTestId } = render(
            <Select label="label" helperText={helperText} items={items} onChange={jest.fn()} />
        );

        expect(getByTestId(/helperText/)).toBeInTheDocument();
        expect(getByText(new RegExp(helperText))).toBeInTheDocument();
    });

    test('Если передать props.onChange, то при выборе item должен вызваться метод onChange', () => {
        const mockOnChange = jest.fn();
        const { getByText, getByTestId } = render(<Select items={items} onChange={mockOnChange} />);

        getByTestId('select_input').focus();

        expect(mockOnChange).not.toHaveBeenCalled();

        fireEvent.click(getByText(new RegExp(items[0].name)));

        expect(mockOnChange).not.toHaveBeenCalledWith(items[0]);
    });

    // eslint-disable-next-line max-len
    test('Если передать props.onChange и props.name, то при выборе item должен вызваться метод onChange, где вторым аргументом будет занчение name', () => {
        const mockOnChange = jest.fn();
        const nockName = 'amazingName';
        const { getByText, getByTestId } = render(
            <Select name={nockName} items={items} onChange={mockOnChange} />
        );

        getByTestId('select_input').focus();

        expect(mockOnChange).not.toHaveBeenCalled();

        const element = getByText(new RegExp(items[0].name));

        fireEvent.mouseDown(element);

        expect(mockOnChange).toHaveBeenCalledWith(items[0], nockName);
    });

    // eslint-disable-next-line max-len
    test('Если открыт выпадающий список, то можно перемещаться клавишами вверх и выбирать по кнопке Enter', () => {
        const mockOnChange = jest.fn();
        const { getByTestId } = render(<Select items={items} onChange={mockOnChange} />);

        getByTestId('select_input').focus();

        expect(mockOnChange).not.toHaveBeenCalled();

        fireEvent.keyDown(getByTestId('select'), { key: 'ArrowUp' });
        fireEvent.keyDown(getByTestId('select'), { key: 'Enter' });

        expect(mockOnChange.mock.calls[0][0]).toEqual(items[1]);
    });

    // eslint-disable-next-line max-len
    test('Если открыт выпадающий список, то можно перемещаться клавишами вниз и выбирать по кнопке "Пробел"', () => {
        const mockOnChange = jest.fn();
        const { getByTestId } = render(<Select items={items} onChange={mockOnChange} />);

        getByTestId('select_input').focus();

        expect(mockOnChange).not.toHaveBeenCalled();

        fireEvent.keyDown(getByTestId('select'), { key: 'ArrowDown' });
        fireEvent.keyDown(getByTestId('select'), { key: ' ' });

        expect(mockOnChange.mock.calls[0][0]).toEqual(items[0]);
    });

    test('Когда в списке выбранный элемент переходит от последнего к первому', () => {
        const { getByTestId, queryByText } = render(<Select items={items} onChange={jest.fn()} />);

        getByTestId('select_input').focus();

        fireEvent.keyDown(getByTestId('select'), { key: 'ArrowDown' });
        fireEvent.keyDown(getByTestId('select'), { key: 'ArrowDown' });
        fireEvent.keyDown(getByTestId('select'), { key: 'ArrowDown' });
        fireEvent.keyDown(getByTestId('select'), { key: ' ' });

        expect(queryByText(/name_1/)).not.toBeInTheDocument();
    });

    // eslint-disable-next-line max-len
    test('Если открыт выпадающий список и нажать Escape, то выпадающий список должен закрыться', () => {
        const { queryAllByRole, getAllByRole, getByTestId } = render(
            <Select items={items} onChange={jest.fn()} />
        );

        getByTestId('select_input').focus();

        expect(getAllByRole('presentation')).toHaveLength(2);

        fireEvent.keyDown(getByTestId('select'), { key: 'Escape' });

        expect(queryAllByRole('presentation')).toHaveLength(0);
    });

    test('Сценарий просмотра placeholder у input-a', () => {
        const { getByTestId } = render(<Select items={items} onChange={jest.fn()} />);

        getByTestId('select_input').focus();

        fireEvent.change(getByTestId('select_input'), { target: { value: 'name_2' } });
        fireEvent.keyDown(getByTestId('select'), { key: 'ArrowDown' });
        fireEvent.keyDown(getByTestId('select'), { key: 'Enter' });

        getByTestId('select_input').focus();

        expect(getByTestId('select')).toMatchSnapshot();
    });
});
