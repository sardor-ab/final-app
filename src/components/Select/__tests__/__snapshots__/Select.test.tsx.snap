// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Снимок компонента 1`] = `
<div
  class=""
  data-test-id="select"
>
  <div>
    <div
      class=""
      data-test-id="select_input-wrapper"
    >
      <input
        autocomplete="off"
        class=""
        data-test-id="select_input"
        id="select"
        placeholder=""
        value=""
      />
    </div>
    <div
      class=""
    >
      <span>
        ‹
      </span>
    </div>
  </div>
</div>
`;

exports[`Select Снимок компонента с props.fixedPosition и видимым выпадающем списком 1`] = `
<div
  class=""
  data-test-id="select"
>
  <div>
    <div
      class=""
      data-test-id="select_input-wrapper"
    >
      <input
        autocomplete="off"
        class=""
        data-test-id="select_input"
        id="select"
        placeholder=""
        value=""
      />
    </div>
    <div
      class=""
    >
      <span>
        ‹
      </span>
    </div>
    <div>
      <ul
        class=""
        style="top: 0px; left: 0px; width: 0px;"
      >
        <li
          class=""
          role="presentation"
        >
          name_1
        </li>
        <li
          class=""
          role="presentation"
        >
          name_2
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Select Снимок компонента с props.fixedPosition=false (position: absolut) и видимым выпадающем списком 1`] = `
<div
  class=""
  data-test-id="select"
>
  <div>
    <div
      class=""
      data-test-id="select_input-wrapper"
    >
      <input
        autocomplete="off"
        class=""
        data-test-id="select_input"
        id="select"
        placeholder=""
        value=""
      />
    </div>
    <div
      class=""
    >
      <span>
        ‹
      </span>
    </div>
    <ul
      class=""
    >
      <li
        class=""
        role="presentation"
      >
        name_1
      </li>
      <li
        class=""
        role="presentation"
      >
        name_2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Select Сценарий просмотра placeholder у input-a 1`] = `
<div
  class=""
  data-test-id="select"
>
  <div>
    <div
      class=""
      data-test-id="select_input-wrapper"
    >
      <input
        autocomplete="off"
        class=""
        data-test-id="select_input"
        id="select"
        placeholder=""
        value=""
      />
    </div>
    <div
      class=""
    >
      <span>
        ‹
      </span>
    </div>
    <ul
      class=""
    >
      <li
        class=""
        role="presentation"
      >
        name_1
      </li>
      <li
        class=""
        role="presentation"
      >
        name_2
      </li>
    </ul>
  </div>
</div>
`;
